src/keystore/SnapKeystore.ts

```diff
@@ -0,0 +1,133 @@
+import {
+  fetcher,
+  conversationReference,
+  keystore,
+  authn,
+  publicKey,
+  signature,
+} from '@xmtp/proto'
+import { Reader, Writer } from 'protobufjs/minimal'
+const { b64Encode, b64Decode } = fetcher
+
+type Codec<T> = {
+  decode(input: Reader | Uint8Array, length?: number): T
+  encode(message: T, writer?: Writer): Writer
+}
+type ApiDefs = {
+  [k: string]: {
+    req: Codec<any> | null
+    res: Codec<any>
+  }
+}
+
+export const defaultSnapOrigin = `local:http://localhost:8080`
+
+export const apiDefs: ApiDefs = {
+  decryptV1: {
+    req: keystore.DecryptV1Request,
+    res: keystore.DecryptResponse,
+  },
+  encryptV1: {
+    req: keystore.EncryptV1Request,
+    res: keystore.EncryptResponse,
+  },
+  encryptV2: {
+    req: keystore.EncryptV2Request,
+    res: keystore.EncryptResponse,
+  },
+  decryptV2: {
+    req: keystore.DecryptV2Request,
+    res: keystore.DecryptResponse,
+  },
+  saveInvites: {
+    req: keystore.SaveInvitesRequest,
+    res: keystore.SaveInvitesResponse,
+  },
+  createInvite: {
+    req: keystore.CreateInviteRequest,
+    res: keystore.CreateInviteResponse,
+  },
+  createAuthToken: {
+    req: keystore.CreateAuthTokenRequest,
+    res: authn.Token,
+  },
+  signDigest: {
+    req: keystore.SignDigestRequest,
+    res: signature.Signature,
+  },
+  getPublicKeyBundle: {
+    req: null,
+    res: publicKey.PublicKeyBundle,
+  },
+} as const
+
+export function SnapKeystore(): Keystore {
+  const generatedMethods: any = {}
+  const ethereum = window.ethereum
+
+  async function ethereumRequest<T extends keyof Keystore>(
+    method: T,
+    req: Uint8Array | null
+  ): Promise<string | string[]> {
+    const response = await ethereum.request({
+      method: 'wallet_invokeSnap',
+      params: {
+        snapId: defaultSnapOrigin,
+        request: {
+          method,
+          params: { req: req ? b64Encode(req, 0, req.length) : null },
+        },
+      },
+    })
+
+    if (!response || typeof response !== 'object') {
+      throw new Error('No response value')
+    }
+
+    return (response as any).res as unknown as string | string[]
+  }
+
+  async function getResponse<T extends keyof Keystore>(
+    method: T,
+    req: Uint8Array | null,
+    resDecoder: typeof apiDefs[T]['res']
+  ): Promise<typeof apiDefs[T]['res']> {
+    const responseString = await ethereumRequest(method, req)
+    if (Array.isArray(responseString)) {
+      throw new Error('Unexpected array response')
+    }
+    return resDecoder.decode(b64Decode(responseString))
+  }
+
+  for (const [method, apiDef] of Object.entries(apiDefs)) {
+    generatedMethods[method] = async (req: any) => {
+      if (!apiDef.req) {
+        return getResponse(method as keyof Keystore, null, apiDef.res)
+      }
+      const reqBytes = apiDef.req.encode(req).finish()
+      return getResponse(method as keyof Keystore, reqBytes, apiDef.res)
+    }
+  }
+
+  return {
+    ...generatedMethods,
+    async getV2Conversations() {
```

Comment:
I knew having the response type be `ConversationReference[]` for this one was going to be annoying, but it bothers me even more than expected. Considering sneaking in a change to v8 to make this consistent with the other APIs and return a single response object with a nested array.